Pause

bradcooper314 opened the pull request:
Merges additions to feature/input as well.

Hacky implementation that does its job. Two issues I am aware of:

1) Pressing the Pause button rapidly shifts from Pause/Unpause as the Input Buffers read if the key is pressed down every frame, not if it's pressed down once. Added this to the backlog and will tackle it later.

2) Had to create a local copy of Listeners in Event Manager. Whenever a State calls Exit/Suspend, it removes its Listener, and likewise adds it whenever it calls Enter/Unsuspend. Doing this calls changes the vector for the Listeners, which in turn cancels all iterators in use. Sadly, this all happens during the dispatch of the event that uses this iterator. Compiler error ahoy!

A local copy solves this problem, but could also be solved by QueueAddListener() and QueueRemoveListener() methods. I took the easier option. Delaying the removal at StateMachine-side doesn't work for the case when Exit() is called, as this cannot be delayed without also keeping that State in the StateMachine stack, which is not ideal.

bradcooper314 committed:
Pause the game. Currently nothing sends the required "Pause" event.

bradcooper314 committed:
Changing process to dispatch events.

bradcooper314 committed:
Merge branch 'feature/input' into pause

bradcooper314 committed:
Changed pause button on keyboard from ESC -> ENTER.
Had to change EventManager because of iterator issue.

MatthewWhite commented on CS483/CS483/Kartaclysm/Data/UserConfig/ControlBindings.xml:
```diff
@@ -9,7 +9,7 @@
       <kartAbility1 value="68" />
       <kartAbility2 value="70" />
-      <pause value="256" />
+      <pause value="257" />
```
Why switch from escape to enter? To me, it seems more intuitive to pause with escape.

bradcooper314 replied:
GLFW is currently accepting ESC as closing the program, so the game would pause and then quit.

MatthewWhite commented on CS483/CS483/Kartaclysm/Services/IO/InputActionMapping.cpp:
```diff
@@ -109,59 +109,70 @@ namespace Kartaclysm
+			// Iterate ability and pause actions
+			if (pBuffer->IsKeyDown(mActionMap[Racer::eDriverAbility1]))
+			{
+				pEventManager->TriggerEvent(new HeatStroke::Event(strPlayerIdentifier + "Driver1"));
```
I would change the end of the string from just "DriverX" to "DriverAbilityX" or something similar to make it more clear what the event is.

bradcooper314 replied:
I can. String comparisons aren't that major I suppose, but I wanted something that was a middle ground between "D1" and "DriverAbility1".

MatthewWhite commented on CS483/CS483/Kartaclysm/StateMachine/Gameplay/StatePaused.cpp:
```diff
@@ -0,0 +1,166 @@
+	p_pEvent->GetOptionalIntParameter("Player", iPlayer, iPlayer);
+
+	// Pop current state
+	printf("Unpausing\n");
```
What's currently stopping you from having a boolean along the lines of m_bPauseInputPressedLastFrame (though a better name would be welcome) that is initially set to true when the state is entered, and set to false when the previous value was true and the pause button is not currently pressed inside StatePaused::Update().  Then check that value here to see if you need to exit.

It seems like a trivial fix, so why are we putting up with the issue of thrashing between states when we know the fix?  If I'm missing something, let me know.

bradcooper314 replied:
It's not just a problem with pausing/unpausing, it's also a problem with abilities at well. It's something that should be handled at the Input level, especially since they already keep track of the previous frame's inputs.

It's a simple method to add to the Input Buffers, ButtonDownOnce() vs ButtonDownContinuous() as examples, maybe ButtonReleaseOnce() if there's a need. I slotted this fix into the Backlog for the next sprint, or at least until I see how long the HUD is going to take me for this sprint.

MatthewWhite reviewed (changes requested):
- Why enter instead of escape?
- Why allow thrashing between states?

MatthewWhite commented on CS483/CS483/Kartaclysm/Services/IO/InputActionMapping.cpp:
```diff
@@ -109,59 +109,70 @@ namespace Kartaclysm
 
-		HeatStroke::Event* pEvent = new HeatStroke::Event("PlayerInput_" + p_iPlayer);
+		// Helpful variables to avoid repetition
+		std::string strPlayerIdentifier = "P" + std::to_string(p_iPlayer) + "_";
```
Why does the playerId have the "_" at the end? Surely it's not relevant to the id, and if the id ever needs to be concatenated with something else, such as "DriverX" for the abilities, it should be the responsibility of the concatenating code to insert the "_".

bradcooper314 replied:
Easy switch. I thought since it was repeated it could be handled just the one time.

bradcooper314 committed:
Created IsKeyDownOnce() method to return true on only the first frame a key is pressed.
Renamed events to be more obvious: "P1_Driver1" becomes "Player1_DriverAbility1".

bradcooper314 committed:
Merge branch 'feature/input' into pause

bradcooper314 committed:
Removed printf lines.

bradcooper314 committed:
Merge branch 'develop' into feature/input

bradcooper314 committed:
Merge branch 'feature/input' into pause
